rust/common/cookieless/src/manager.rs

```diff
@@ -0,0 +1,895 @@
+    ChronoError(#[from] chrono::ParseError),
+
+    #[error("Timezone error: {0}")]
+    TimezoneError(String),
+
+    #[error("Invalid identify count: {0}")]
+    InvalidIdentifyCount(String),
+
+    #[error("Redis error: {0}")]
+    RedisError(String),
+}
+
+/// Configuration for the CookielessManager
+#[derive(Debug, Clone)]
+pub struct CookielessConfig {
+    /// Whether cookieless mode is disabled
+    pub disabled: bool,
+    /// Whether to force stateless mode
+    pub force_stateless_mode: bool,
+    /// TTL for identifies (in seconds)
+    pub identifies_ttl_seconds: u64,
+    /// TTL for sessions (in seconds)
+    pub session_ttl_seconds: u64,
+    /// TTL for salts (in seconds)
+    pub salt_ttl_seconds: u64,
+    /// Session inactivity timeout (in milliseconds)
+    pub session_inactivity_ms: u64,
+}
+
+impl Default for CookielessConfig {
+    fn default() -> Self {
+        Self {
+            disabled: false,
+            force_stateless_mode: false,
+            identifies_ttl_seconds: IDENTIFIES_TTL_SECONDS,
+            session_ttl_seconds: SESSION_TTL_SECONDS,
+            salt_ttl_seconds: SALT_TTL_SECONDS,
+            session_inactivity_ms: SESSION_INACTIVITY_MS,
+        }
+    }
+}
+
+/// Parameters for computing a hash
+#[derive(Debug, Clone)]
+pub struct HashParams<'a> {
+    /// Timestamp in milliseconds
+    pub timestamp_ms: u64,
+    /// Event timezone
+    pub event_time_zone: Option<&'a str>,
+    /// Team timezone
+    pub team_time_zone: &'a str,
+    /// Team ID
+    pub team_id: u64,
+    /// IP address
+    pub ip: &'a str,
+    /// Host
+    pub host: &'a str,
+    /// User agent
+    pub user_agent: &'a str,
+    /// Counter value
+    pub n: u64,
+    /// Additional data to include in the hash
+    pub hash_extra: &'a str,
+}
+
+/// Data for an event to be processed by the cookieless manager
+#[derive(Debug, Clone)]
+pub struct EventData<'a> {
+    /// IP address
+    pub ip: &'a str,
+    /// Timestamp in milliseconds
+    pub timestamp_ms: u64,
+    /// Host
+    pub host: &'a str,
+    /// User agent
+    pub user_agent: &'a str,
+    /// Event timezone (optional)
+    pub event_time_zone: Option<&'a str>,
+    /// Additional data to include in the hash (optional)
+    pub hash_extra: Option<&'a str>,
+    /// Team ID
+    pub team_id: u64,
+    /// Team timezone (optional)
+    pub team_time_zone: Option<&'a str>,
+}
+
+/// Manager for cookieless tracking
+pub struct CookielessManager {
+    /// Configuration for the manager
+    pub config: CookielessConfig,
+    /// Salt cache for retrieving and storing salts
+    salt_cache: SaltCache,
+    /// Redis client for direct access
+    redis_client: Arc<dyn RedisClient + Send + Sync>,
+}
+
+impl CookielessManager {
+    /// Create a new CookielessManager
+    pub fn new(config: CookielessConfig, redis_client: Arc<dyn RedisClient + Send + Sync>) -> Self {
+        let salt_cache = SaltCache::new(redis_client.clone(), Some(config.salt_ttl_seconds));
```

Comment:
Is cloning the client fine? I'm worried it might create a new unnecessary connection. Because this is an Arc, doesn't that imply we don't need copies, people can share its reference?

Or I might be a newbie, and calling `clone` on an `Arc` is what lets us actually share it.

Reply:
> Or I might be a newbie, and calling clone on an Arc is what lets us actually share it.

Yeah exactly, it's not creating a new redis connection, it's just creating a new pointer to it